File: python/data_structures/binary_heap.py
"""
Created by Connor Murray (connormurray7@gmail.com)
8/27/2016

Python implementation of a binary heap.
"""

class BinaryHeap:
    """Binary max heap, array as underlying data structure

    Attributes:
       data: list as the underlying data structure
       size: current length of the list
       max_size: maximum number of vals that can fit
       comp: comparator function that dictates the heap ordering property
    """

    def __init__(self, init_size=1, comp=None):
        self.heap = init_size*[None]
        self.size = 0
        if comp is None:
            self.comp = lambda x, y: x > y
        else:
            self.comp = comp

    def __str__(self):
        """String representation of the heap"""
        vals = []
        for val in self.heap:
            if val is not None:
                vals.append(val)
        out_str = ' '.join([str(s) for s in vals])
        return "Binary Heap [" + out_str + "]"

    def fix_up(self, idx):
        """Sorts upwards in heap, runs in O(log n). """
        parent = idx // 2
        while parent:
            if self.comp(self.heap[idx], self.heap[parent]):
                self.swap(idx, parent)
            idx = parent
            parent = parent // 2

    def fix_down(self, idx):
        """Sorts downward in heap, runs in O(log n)."""
        cur = idx
        child = self.left(idx)
        while child <= self.size:
            child = self.min_child(cur)
            if self.comp(self.heap[child], self.heap[cur]):
                self.swap(cur, child)
            cur = child
            child = self.left(child)

    def min_child(self, idx):
        """Returns the index of the minimum child"""
        left = self.left(idx)
        right = self.right(idx)
        if right > self.size:
            return left
        elif self.comp(self.heap[left], self.heap[right]):
            return left
        return right

    def build_heap(self, array):
        """Constructs heap from given array, O(n log n)"""
        self.size = len(array)
        self.heap = array
        self.heap.insert(0, None)
        for i in range(len(self.heap)//2, 0, -1):
            self.fix_down(i)

    def peek(self):
        """Returns the top of the heap, O(1)."""
        return self.heap[1]

    def pop(self):
        """Returns the top of the heap and removes it, O(log n)."""
        if self.size == 0:
            return None
        val = self.heap[1]
        self.swap(1, self.size)
        self.heap.pop()
        self.size -= 1
        self.fix_down(1)
        return val

    def push(self, val):
        """Puts value into heap, runs in O(log n)."""
        self.size += 1
        self.heap.append(val)
        self.fix_up(self.size)

    @staticmethod
    def left(idx):
        """Returns index of left child"""
        return idx * 2

    @staticmethod
    def right(idx):
        """Returns index of right child"""
        return idx * 2 + 1

    def swap(self, i, j):
        """Swaps vals in heap """
        tmp = self.heap[i]
        self.heap[i] = self.heap[j]
        self.heap[j] = tmp

File: cpp/data_structures/tree/binary_heap/BinaryHeap.cpp
#include "BinaryHeap.h"
#include <memory>
#include <utility>


template<typename K, typename V, typename C>
BinaryHeap<K,V,C>::BinaryHeap() {
    heap = std::vector<std::unique_ptr<TreeNode<K,V>>>();
    heap.push_back(nullptr); //Not using index 0
    comp = C();
}

template<typename K, typename V, typename C>
BinaryHeap<K,V,C>::BinaryHeap(const BinaryHeap& other) {
    heap = other.heap;
    comp = other.comp;
}

template<typename K, typename V, typename C>
BinaryHeap<K,V,C>::~BinaryHeap() {}

template<typename K, typename V, typename C>
void BinaryHeap<K,V,C>::push(const K& key, const V& val) {
    heap.push_back(std::make_unique<TreeNode<K,V>>(key, val));
    fix_up(heap.size()-1);
}

template<typename K, typename V, typename C>
void BinaryHeap<K,V,C>::pop() {
    std::swap(heap[1], heap[heap.size()-1]);
    heap.pop_back();
    fix_down(1);
}

template<typename K, typename V, typename C>
V& BinaryHeap<K,V,C>::top() const {
    return heap[1]->value;
}

template<typename K, typename V, typename C>
bool BinaryHeap<K,V,C>::empty() const {
    return heap.size() == 1; //First entry is not used
}

template<typename K, typename V, typename C>
void BinaryHeap<K,V,C>::fix_down(int node_idx) {
    unsigned int size = heap.size()-1;
    unsigned int parent = node_idx;

    while(2*parent <= size) {
        unsigned int child = 2 * parent; //left first
        if(child < size && comp(heap[child]->key, heap[child+1]->key)) {
            ++child; //right child
        }
        if(heap[child]->key < heap[parent]->key) {
            break;
        }
        std::swap(heap[child], heap[parent]);
        parent = child;
    }
}

template<typename K, typename V, typename C>
void BinaryHeap<K,V,C>::fix_up(int node) {
    int cur_node = node;
    while(cur_node > 1 && comp(heap[cur_node/2]->key, heap[cur_node]->key)) {
        std::swap(heap[cur_node], heap[cur_node/2]);
        cur_node /= 2;
    }
}
